{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response Figure 1\n",
    "\n",
    "Since this manuscript's main advance regarding variation in paternal age effect over the Rahbari, et al. result is greater statistical power, more robust statistical analyses of this pattern would strengthen the paper. Figure 3 presents a commendable amount of raw data in a fairly clear way, yet the authors use only a simple ANOVA to test whether different families have different dependencies on paternal age. The supplement claims that this result cannot be an artifact of low sequencing coverage because regions covered by <12 reads are excluded from the denominator, but there still might be subtle differences in variant discovery power between e.g. regions covered by 12 reads and regions covered by 30 reads. To hedge against this, the authors can define the \"callable genome\" continuously (point 1 under \"other questions and suggestions\") or they can check whether mean read coverage appears to covary with mutation rates across individuals after filtering away the regions covered by <12 reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "\n",
    "# read in second- and third-generation DNMs (now with a column for mean read depth)\n",
    "gen2 = read.csv(\"../data/second_gen.dnms.summary.csv\")\n",
    "gen3 = read.csv(\"../data/third_gen.dnms.summary.csv\")\n",
    "\n",
    "gen2$generation = \"2nd\"\n",
    "gen3$generation = \"3rd\"\n",
    "\n",
    "# combine the second- and third-generation dataframes, and calculate\n",
    "# autosomal mutation rates for all samples\n",
    "combined = rbind(gen2, gen3)\n",
    "combined$autosomal_mutation_rate = combined$autosomal_dnms / (combined$autosomal_callable_fraction * 2)\n",
    "\n",
    "# generate the figure\n",
    "p <- ggplot(combined, aes(x=autosomal_mutation_rate, y=mean_depth)) +\n",
    "        facet_wrap(~generation) +\n",
    "        geom_smooth(method=\"lm\", aes(col=generation)) +\n",
    "        geom_point(aes(fill=generation), pch=21, col=\"white\", size=2) +\n",
    "        xlab(\"Autosomal mutation rate\") +\n",
    "        ylab(\"Mean autosomal read depth\\n(sites >= 12 reads)\") +\n",
    "        theme(axis.text.x = element_text(angle=45, vjust=0.5)) +\n",
    "        coord_fixed(ratio=9.5e-10)\n",
    "\n",
    "# fit models predicting mean autosomal read depth as a function\n",
    "# of the autosomal mutation rate\n",
    "m_second_gen = lm(mean_depth ~ autosomal_mutation_rate, data=subset(combined, generation==\"2nd\"))\n",
    "m_third_gen = lm(mean_depth ~ autosomal_mutation_rate, data=subset(combined, generation==\"3rd\"))\n",
    "\n",
    "# test whether read depth and mutation rates are correlated\n",
    "summary(m_second_gen)\n",
    "summary(m_third_gen)\n",
    "\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response Figure 2\n",
    "\n",
    "Another concern about paternal age effects is the extent to which outlier offspring may be driving the apparent rate variation across families. If the authors were to randomly sample half of the children from each family and run the analysis again, how much is the paternal age effect rank preserved? Alternatively, how much is the family rank ordering preserved if mutations are only called from a subset of the chromosomes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(MASS)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "library(ggridges)\n",
    "library(tidyr)\n",
    "library(reshape2)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "\n",
    "set.seed(12345)\n",
    "\n",
    "# read in third-generation DNMs\n",
    "gen3 = read.csv(\"../../dnm-analysis/ceph/manuscript/third_gen.dnms.summary.csv\")\n",
    "\n",
    "# number of subsampling experiments to perform\n",
    "n_sims = 100\n",
    "\n",
    "# empty dataframe to store ranks from subsampling experiments\n",
    "sub_df = data.frame(matrix(NA, nrow=n_sims*40, ncol=2))\n",
    "# empty dataframe to store original ranks\n",
    "orig_df = data.frame()\n",
    "\n",
    "fam_list = unique(gen3$family_id)\n",
    "\n",
    "# this function accepts a dataframe with information\n",
    "# about each family's slope, and sorts it in order of increasing\n",
    "# slope\n",
    "sort_and_index <- function(df) {\n",
    "    df = df[order(df$slope),]\n",
    "    df$facet_order = factor(df$family_id, levels = unique(df$family_id))\n",
    "    df = df[!duplicated(df[,c('family_id')]),]\n",
    "    df$order = rev(c(1:nrow(df)))\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "# loop over every family, fit a regression (using all of the\n",
    "# third-generation samples in the family), and store the results\n",
    "# in `orig_df`\n",
    "for (sp in split(gen3, as.factor(gen3$family_id))) {\n",
    "    m = glm(autosomal_dnms ~ dad_age, data=sp, family=poisson(link=\"identity\"))\n",
    "    s = summary(m)\n",
    "    sp$slope = s$coefficients[[2]]\n",
    "    sp$intercept = s$coefficients[[1]]\n",
    "    orig_df = rbind(orig_df, sp)\n",
    "}\n",
    "\n",
    "# sort and rank results from the full third-generation dataset\n",
    "orig_df = sort_and_index(orig_df)\n",
    "orig_rank_order = rev(c(as.character(orig_df$family_id)))\n",
    "\n",
    "# for the specified number of simulations, subsample each\n",
    "# family's children, fit a regression, and store the results in\n",
    "# `sub_df`\n",
    "for (iter in 1:n_sims) {\n",
    "    new_gen3 = data.frame()\n",
    "    for (sp in split(gen3, as.factor(gen3$family_id))) {\n",
    "        # sub sample the family's children (here, 75% of children)\n",
    "        sub_sampled = sp[sample(1:nrow(sp), nrow(sp) * 0.75, replace=FALSE),] \n",
    "        m = glm(autosomal_dnms ~ dad_age, data=sub_sampled, family=poisson(link=\"identity\"))\n",
    "        s = summary(m)\n",
    "        sub_sampled$slope = s$coefficients[[2]]\n",
    "        new_gen3 = rbind(new_gen3, sub_sampled)\n",
    "    }\n",
    "    \n",
    "    # sort and rank results\n",
    "    sorted_df = sort_and_index(new_gen3)\n",
    "    \n",
    "    # add the results to the main `sub_df`\n",
    "    for (f in 1:length(fam_list)) {\n",
    "        # get the index (rank) of each family ID in `sorted_df`,\n",
    "        # which contains the new rank of each family ID after subsampling\n",
    "        sub_df[iter * f,2] = which(fam_list[[f]] == sorted_df$family_id)\n",
    "        sub_df[iter * f,1] = as.character(fam_list[[f]])\n",
    "    }\n",
    "}\n",
    "\n",
    "colnames(sub_df) = c(\"family_id\", \"rank\")\n",
    "\n",
    "# make sure `family_id` is treated as a factor\n",
    "sub_df$family_id <- as.factor(sub_df$family_id)\n",
    "\n",
    "# create a \"ridges\" or \"joyplot\" summarizing the\n",
    "# distribution of ranks for each family across \n",
    "# simulations, ordered by their original ranks\n",
    "p <- ggplot(sub_df, aes(x = rank, y = family_id, fill=..x..)) + \n",
    "     geom_density_ridges_gradient(scale = 3, rel_min_height = 0.01, quantile_lines = TRUE, quantiles = 2) +\n",
    "     scale_fill_gradient(name = \"Rank\", low=\"dodgerblue\", high=\"firebrick\") +\n",
    "     scale_y_discrete(limits=orig_rank_order) +\n",
    "     xlab(\"Distribution of family ranks\\nfollowing 100 sampling trials\") +\n",
    "     ylab(\"Family ID (original rankings)\") +\n",
    "     theme(axis.text.x = element_text(size=12)) +\n",
    "     theme(axis.text.y = element_text(size=10))\n",
    "\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response Figure 3\n",
    "\n",
    "A factor regarding paternal age effects that only mentioned briefly alluded to late in the paper are the differences in the intercept between families (Figure 3c). Do either the intercept or slope vary with the number of F2 children? Is there any (anti-)correlation between slope and intercept? It would seem odd if the intercept strongly impacts the slope since a low per-year rate probably should not correlate with a high initial rate at younger ages."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Is there anti-correlation between slopes and intercepts in CEPH families?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the correlation between family size and intercept\n",
    "p <- ggplot(orig_df, aes(x=slope, y=intercept)) +\n",
    "        geom_smooth(method='lm', color='firebrick', alpha=0.25) +\n",
    "        geom_point(pch=21, fill=\"black\", col=\"white\", size=3, lwd=0.5) +\n",
    "        ylab('Initial mutation count (intercept) in family') +\n",
    "        xlab('Slope in family') +\n",
    "        theme(axis.text.x=element_text(size=16)) +\n",
    "        theme(axis.text.y=element_text(size=16)) +\n",
    "        theme(text=element_text(size=16))\n",
    "\n",
    "m = lm(intercept ~ slope, data=orig_df)\n",
    "summary(m)\n",
    "\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Anti-correlation between slopes and intercepts is expected for randomly distributed DNM counts..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(MASS)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "\n",
    "# set seed so that example is reproducible\n",
    "set.seed(1234)\n",
    "\n",
    "gen3 = read.csv(\"../data/third_gen.dnms.summary.csv\")\n",
    "\n",
    "library(dplyr)\n",
    "\n",
    "df1 = gen3[,c(\"dad_age\",\"family_id\")]\n",
    "FUN <- function(x) rpois(lambda=x * 1.72 + 15, n=1)\n",
    "\n",
    "# randomly assign DNM counts to each third-generation sample based\n",
    "# on their paternal age at birth\n",
    "df1$dnms = lapply(df1$dad_age, FUN)\n",
    "df1$dnms = as.numeric(df1$dnms)\n",
    "\n",
    "colnames(df1) = c('age', 'fam_id', 'dnms')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the slopes for each family and add\n",
    "# to a new dataframe\n",
    "new_df1 = data.frame()\n",
    "for (sp in split(df1, df1$fam_id)) {\n",
    "    m = glm(dnms ~ age, data=sp, family=poisson(link=\"identity\"))\n",
    "    s = summary(m)\n",
    "    sp$slope = s$coefficients[[2]]\n",
    "    sp$intercept = s$coefficients[[1]]\n",
    "    new_df1 = rbind(new_df1, sp)\n",
    "}\n",
    "\n",
    "# get rid of duplicates (i.e., samples from the same\n",
    "# family)\n",
    "sort_and_index <- function(df) {\n",
    "    df = df[order(df$slope),]\n",
    "    df$facet_order = factor(df$fam_id, levels = unique(df$fam_id))\n",
    "    df = df[!duplicated(df[,c('fam_id')]),]\n",
    "    df$order = rev(c(1:nrow(df)))\n",
    "    return(df)\n",
    "}\n",
    "    \n",
    "new_df1 = sort_and_index(new_df1)\n",
    "\n",
    "# plot slopes and intercepts\n",
    "p <- ggplot(new_df1, aes(x=slope, y=intercept)) +\n",
    "        geom_smooth(method='lm', color='dodgerblue', alpha=0.25) +\n",
    "        geom_point(pch=21, fill=\"black\", col=\"white\", size=3, lwd=0.5) +\n",
    "        ylab('Initial mutation count (intercept) in family') +\n",
    "        xlab('Slope in family') +\n",
    "        theme(axis.text.x=element_text(size=16)) +\n",
    "        theme(axis.text.y=element_text(size=16)) +\n",
    "        theme(text=element_text(size=16))\n",
    "p\n",
    "\n",
    "print(cor.test(new_df1$slope, new_df1$intercept))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ...but inter-family variability is not expected for randomly distributed DNM counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = glm(dnms ~ age * fam_id, data=df1, family=poisson(link=\"identity\"))\n",
    "anova(m, test=\"Chisq\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Is family size correlated with the slope or intercept for a family?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(MASS)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "\n",
    "# read in third-generation DNMs\n",
    "gen3 = read.csv(\"../../dnm-analysis/ceph/manuscript/third_gen.dnms.summary.csv\")\n",
    "\n",
    "orig_df = data.frame()\n",
    "\n",
    "# this function accepts a dataframe with information\n",
    "# about each family's slope, and sorts it in order of increasing\n",
    "# slope\n",
    "sort_and_index <- function(df) {\n",
    "    df = df[order(df$slope),]\n",
    "    df$facet_order = factor(df$family_id, levels = unique(df$family_id))\n",
    "    df = df[!duplicated(df[,c('family_id')]),]\n",
    "    df$order = rev(c(1:nrow(df)))\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "# loop over every family, fit a regression, and store the results\n",
    "# in `orig_df`\n",
    "for (sp in split(gen3, as.factor(gen3$family_id))) {\n",
    "    m = glm(autosomal_dnms ~ dad_age, data=sp, family=poisson(link=\"identity\"))\n",
    "    s = summary(m)\n",
    "    sp$slope = s$coefficients[[2]]\n",
    "    sp$intercept = s$coefficients[[1]]\n",
    "    orig_df = rbind(orig_df, sp)\n",
    "}\n",
    "\n",
    "# sort the rank results from the full third-generation dataset\n",
    "orig_df = sort_and_index(orig_df)\n",
    "\n",
    "# plot the correlation between family size and slope\n",
    "p1 <- ggplot(orig_df, aes(x=n_sibs, y=slope)) +\n",
    "        geom_smooth(method='lm', color='firebrick', alpha=0.25) +\n",
    "        geom_point(pch=21, fill=\"black\", col=\"white\", size=4, lwd=0.5) +\n",
    "        ylab('Paternal age effect (slope) in family') +\n",
    "        xlab('Number of siblings in family') +\n",
    "        theme(axis.text.x=element_text(size=16)) +\n",
    "        theme(axis.text.y=element_text(size=16)) +\n",
    "        theme(text=element_text(size=16))\n",
    "\n",
    "\n",
    "m = lm(slope ~ n_sibs, data=orig_df)\n",
    "summary(m)\n",
    "\n",
    "# plot the correlation between family size and intercept\n",
    "p2 <- ggplot(orig_df, aes(x=n_sibs, y=intercept)) +\n",
    "        geom_smooth(method='lm', color='firebrick', alpha=0.25) +\n",
    "        geom_point(pch=21, fill=\"black\", col=\"white\", size=4, lwd=0.5) +\n",
    "        ylab('Initial mutation count (intercept) in family') +\n",
    "        xlab('Number of siblings in family') +\n",
    "        theme(axis.text.x=element_text(size=16)) +\n",
    "        theme(axis.text.y=element_text(size=16)) +\n",
    "        theme(text=element_text(size=16))\n",
    "\n",
    "\n",
    "m = lm(intercept ~ n_sibs, data=orig_df)\n",
    "summary(m)\n",
    "\n",
    "p1\n",
    "p2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other questions and suggestions "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Suggestion #5. \n",
    "\n",
    "Other things to explore further related to parental age effects are: how do the conclusions change and/or can you detect similar variability in maternal age when analyzing phased DNMs? This may be underpowered, but for those families that share grandparents, if two brothers are in the F1 generation, do their paternal age effects differ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen3 = read.csv('../data/third_gen.dnms.summary.csv')\n",
    "\n",
    "# fit model with only paternally-phased counts\n",
    "m = glm(dad_dnms_auto ~ dad_age * family_id, data=gen3, family=poisson(link=\"identity\"))\n",
    "\n",
    "anova(m, test=\"Chisq\")\n",
    "\n",
    "# fit model with only maternally-phased counts.\n",
    "# need to add a pseudo-count of 1 to maternal DNMs first.\n",
    "gen3$mom_dnms_auto = gen3$mom_dnms_auto + 1\n",
    "\n",
    "m = glm(mom_dnms_auto ~ mom_age * family_id, data=gen3, family=poisson(link=\"identity\"))\n",
    "\n",
    "anova(m, test=\"Chisq\")\n",
    "\n",
    "# identify difference in paternal age effects between pairs of brothers\n",
    "\n",
    "# first, for family 24\n",
    "gen3_fam24 = subset(gen3, family_id %in% c(\"24_C\", \"24_D\"))\n",
    "\n",
    "m_exp = glm(autosomal_dnms ~ dad_age * family_id, data=gen3_fam24, family=poisson(link=\"identity\"))\n",
    "m_null = glm(autosomal_dnms ~ dad_age, data=gen3_fam24, family=poisson(link=\"identity\"))\n",
    "anova(m_null, m_exp, test=\"Chisq\")\n",
    "\n",
    "\n",
    "# next, for family 19\n",
    "gen3_fam19 = subset(gen3, family_id %in% c(\"19_A\", \"19_B\"))\n",
    "\n",
    "m_exp = glm(autosomal_dnms ~ dad_age * family_id, data=gen3_fam19, family=poisson(link=\"identity\"))\n",
    "m_null = glm(autosomal_dnms ~ dad_age, data=gen3_fam19, family=poisson(link=\"identity\"))\n",
    "anova(m_null, m_exp, test=\"Chisq\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
